app/scripts/lib/operation-safener.ts

```diff
@@ -0,0 +1,137 @@
+   */
+  op: O;
+  /**
+   * The wait time in milliseconds for the debounce function. Defaults to 0ms.
+   */
+  wait?: number;
+  /**
+   * Additional options for the debounce function. See
+   * [lodash's debounce documentation](https://lodash.com/docs/4.17.15#debounce)
+   * for more details.
+   */
+  options?: DebounceSettings;
+};
+
+export class OperationSafener<O extends Op = Op> {
+  /**
+   * A debounced function that wraps the operation to be executed.
+   * It ensures that the operation is not called too frequently. And that no
+   * two operations are ever executed at the same time.
+   */
+  #bouncer: DebouncedFunc<(...params: Parameters<O>) => Promise<ReturnType<O>>>;
+
+  #evacuating: Promise<void> | null = null;
+
+  /**
+   * Creates an instance of OperationSafener.
+   *
+   * This class is designed to safely execute operations that may be
+   * invoked frequently or in quick succession. It debounces the execution of
+   * the operation, ensuring that it is only executed once within a specified
+   * wait time.
+   *
+   * Additionally, it ensures that no two operations are executed
+   * at the same time, if any operation might take longer than the `wait` time
+   * to complete.
+   *
+   * Finally, it provides a method to evacuate the current operation queue,
+   * executing the latest pending operation and preventing any future
+   * operations from being queued.
+   *
+   * @param config - Configuration object for the OperationSafener.
+   * @param config.op
+   * @param config.wait
+   * @param config.options
+   * @example
+   * ```ts
+   * const operationSafener = new OperationSafener({
+   *   op: async (data) => {
+   *     // Perform some operation with data
+   *     console.log('Operation executed with:', data);
+   *   },
+   *   wait: 1000, // Wait for 1 second before executing the operation
+   * });
+   * // queued
+   * operationSafener.execute('example data 1');
+   * // queued, previous call is removed
+   * operationSafener.execute('example data 2');
+   * // _immediately_ logs `example data 2` _only_
+   * await operationSafener.evacuate();
+   * // never logs, call to `evacuate` prevents any future operations.
+   * operationSafener.execute('example data 3');
+   * ```
+   */
+  constructor({ op, wait, options }: Config<O>) {
+    const opts: LockOptions = { mode: 'exclusive' };
+    const func = (...params: Parameters<O>) =>
+      navigator.locks.request('lock', opts, () => op(...params));
+    this.#bouncer = debounce(func, wait, options);
+  }
+
+  /**
+   * Evacuates the current operation queue, executing the latest pending
+   * operation and preventing any future operations from being queued.
+   *
+   * @returns A Promise that resolves when the evacuation is complete.
+   */
+  evacuate = () => {
+    if (this.#evacuating) {
+      log.warn('already evacuating');
+      return this.#evacuating;
+    }
+
+    // eslint-disable-next-line no-async-promise-executor
```

Comment:
Nit: Here's my attempt to explain why this isn't a mistake
```suggestion
    // Intentionally using async executor to ensure `this.#evacuating` never rejects
    // If `flush` throws an error, it will be captured by Sentry as an unhandled Promise rejection
    // eslint-disable-next-line no-async-promise-executor
```

Reply:
Sorry I meant the reason for using an async executor specifically, not the purpose for the `#evacuate` property. The ESLint warning wouldn't complain if the executor was synchronous.

Reply:
We just need to wait for the debounced flush operation to finalize before resolving/returning from evacuate. And we need all calls to `evacuate` to wait for that same operation. I'll rewrite to get rid of the eslint disable line and hopefully clear up the error handling confusion.